bdai_ros2_wrappers/bdai_ros2_wrappers/action.py

```diff
@@ -0,0 +1,583 @@
+    and asynchronous action invocation, in a way that resembles remote procedure calls.
+    """
+
+    def __init__(self, action_type: Type, action_name: str, node: Optional[Node] = None, **kwargs: Any) -> None:
+        """
+        Initializes the actionable.
+
+        Args:
+            action_type: Target action type class (as generated by ``rosidl``).
+            action_name: Name of the target action in the ROS 2 graph.
+            node: optional node for the underlying action client, defaults to
+            the current process node.
+
+        All other keyword arguments are forward to the underlying action client.
+        See `rclpy.action.ActionClient` documentation for further reference.
+        """
+        node = node or scope.node()
+        self._action_type = action_type
+        self._action_name = action_name
+        self._action_client = ActionClient(node, action_type, action_name, **kwargs)
+
+    @property
+    def action_name(self) -> str:
+        """Get the target action name."""
+        return self._action_name
+
+    @property
+    def action_type(self) -> Type:
+        """Get the target action type."""
+        return self._action_type
+
+    @property
+    def action_client(self) -> ActionClient:
+        """Get the underlying action client."""
+        return self._action_client
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        """Forward invocation to `Actionable.synchronously`."""
+        return self.synchronously(*args, **kwargs)
+
+    def synchronously(
+        self, goal: Any, *, feedback_callback: Optional[Callable] = None, timeout_sec: Optional[float] = None
+    ) -> Any:
+        """
+        Invoke action synchronously.
+
+        Args:
+            goal: goal to invoke action with.
+            feedback_callback: optional action feedback callback.
+            timeout_sec: optional action timeout, in seconds. If a timeout is specified and it
+            expires, the action will be cancelled and the call will raise. Note this timeout is
+            local to the caller. It may take some time for the action to be cancelled, that is
+            if cancellation is not rejected by the action server.
+
+        Returns:
+            the action result.
+
+        Raises:
+            ActionTimeout: if the action timed out.
+            ActionRejected: if the action was not accepted.
+            ActionCancelled: if the action was cancelled.
+            ActionAborted: if the action was aborted.
+        """
+        action = ActionFuture(self._action_client.send_goal_async(goal, feedback_callback))
+        if not wait_for_future(action.finalization, timeout_sec=timeout_sec):
+            action.cancel()  # proactively cancel
+            raise ActionTimeout()
+        if not action.accepted:
+            raise ActionRejected()
+        if action.cancelled:
+            raise ActionCancelled()
+        if action.aborted:
+            raise ActionAborted()
```

Comment:
Something that has been discussed is there is information that can be sent back as a result even when an action is aborted. This could be failure reasons and information that could be used for error recovery.

Reply:
Unfortunately only user-defined action results are sent back, by design. I could tuck the result received on abortion in the `ActionAborted` exception, but for that to be informative we have to make sure our action result can carry that information and that we populate it.

Reply:
As an afterthought, RTTI is a given in Python. We could define some convention of ours, and then have these wrappers check if some action follows the convention and pay attention to it if it does.

Reply:
I think it's worth at least looking into defining our own error handling convention. Could help a lot with debugging stuff in the future too

Reply:
SGTM. I would defer that design and implementation to follow-up PRs, if everyone agrees.

Reply:
Sure. We should have that discussion though because I believe it will affect the `SpotBasic` refactor.

Reply:
TIcket opened. Deferring.